contracts/smart-account/modules/AccountRecoveryModule.sol

```diff
@@ -0,0 +1,580 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import {BaseAuthorizationModule} from "./BaseAuthorizationModule.sol";
+import {UserOperation} from "@account-abstraction/contracts/interfaces/UserOperation.sol";
+import {ECDSA} from "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+
+/**
+ * @title Account Recovery module for Biconomy Smart Accounts.
+ * @dev Compatible with Biconomy Modular Interface v 0.1
+ *         - It allows the submission and execution of recovery requests
+ *         - EOA guardians only
+ *         - For security reasons the guardian's address is not stored,
+ *           instead, its signature over CONTROL_HASH is used
+ *         - Security delay is always applied to new guardians and recovery requests
+ *         - It is highly recommended to not set security delay to 0
+ * @dev For the validation stage (validateUserOp) can not use custom errors
+ * as EntryPoint contract doesn't support custom error messages within its
+ * 'try IAccount(sender).validateUserOp catch Error' expression
+ * so it becomes more difficult to debug validateUserOp if it uses custom errors
+ * For the execution methods custom errors are used
+ *
+ * @author Fil Makarov - <filipp.makarov@biconomy.io>
+ * based on https://vitalik.ca/general/2021/01/11/recovery.html by Vitalik Buterin
+ */
+
+contract AccountRecoveryModule is BaseAuthorizationModule {
+    using ECDSA for bytes32;
+
+    struct TimeFrame {
+        uint48 validUntil;
+        uint48 validAfter;
+    }
+
+    struct SaSettings {
+        uint8 guardiansCount;
+        uint8 recoveryThreshold;
+        uint48 securityDelay;
+    }
+
+    struct RecoveryRequest {
+        bytes32 callDataHash;
+        uint48 requestTimestamp;
+    }
+
+    string public constant NAME = "Account Recovery Module";
+    string public constant VERSION = "0.1.0";
+
+    bytes32 public constant CONTROL_HASH =
```

Comment:
I would prefer to store it like 0x33e4b41198cc5b8053630ed667ea7c0c4c873f7fc8d9a478b5d7259cec0a4a00 directly what kecckak value it is. and mention in comment above like it is actually calculated like this...

// keccak256(abi.encodePacked("ACCOUNT RECOVERY GUARDIAN SECURE MESSAGE"));

Reply:
fixed